File: packages/python/paybot_sdk/mpp_seam.py
"""Thin MPP (Machine Payments Protocol — Stripe/Tempo) capability seam.

Python port of ``src/mpp-seam.ts``. This is deliberately NOT a client.

WHY a seam and not a client: at time of writing, Stripe/Tempo MPP is still in
PREVIEW — version-pinned, bound to the Tempo L1, and sharing NO signing code with
our EIP-3009 core (MPP uses a different challenge/credential model). Building a
full MPP client now would mean committing to a moving, preview-stage target with a
separate crypto stack.

Instead this module provides the *shape* of MPP support: it can DETECT whether a
server advertises MPP (so a caller can negotiate/route), and exposes a ``settle``
entry point that LOUDLY raises ``MPP_NOT_IMPLEMENTED`` (HTTP 501). When MPP exits
preview, the real settlement path slots in behind this seam without changing the
public surface. Full MPP is deferred to GA.

Dependencies: ``paybot_sdk.errors`` (PayBotApiError).
"""
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any, Callable, Dict, Literal, Optional

from .errors import PayBotApiError


@dataclass
class MppCapability:
    """The result of probing a server for MPP support.

    :param supported: Whether any MPP advertisement was detected.
    :param mode: ``'none'`` | ``'detect-only'`` | ``'preview'`` (today only the
        first two are produced).
    :param spec_version: The advertised spec/version string, when provided.
    :param reason: Human-readable reason, populated when ``supported`` is ``False``.
    """

    supported: bool
    mode: Literal["none", "detect-only", "preview"]
    spec_version: Optional[str] = None
    reason: Optional[str] = None


# Header names that, when present, indicate a server advertises MPP. Compared
# case-insensitively.
_MPP_VERSION_HEADERS = ("mpp-version", "stripe-version")


def _read_header_ci(headers: Dict[str, str], name: str) -> Optional[str]:
    """Read a header value by name, case-insensitively. ``None`` when absent."""
    target = name.lower()
    for key, value in headers.items():
        if key.lower() == target:
            return value
    return None


def detect_mpp_capability(
    server_headers: Optional[Dict[str, str]] = None,
) -> MppCapability:
    """Detect whether response headers advertise MPP support.

    Detection is purely passive — this NEVER signs or settles. It inspects, case
    insensitively, for either an ``MPP-Version`` / ``Stripe-Version`` header
    (value becomes ``spec_version``) or a ``WWW-Authenticate`` challenge whose
    value mentions ``Payment`` (an MPP-style 402 challenge).

    :param server_headers: Response headers to inspect. ``None``/empty → not supported.
    :returns: An :class:`MppCapability`.

    :example:
        >>> detect_mpp_capability({"MPP-Version": "2024-09-preview"}).supported
        True
        >>> detect_mpp_capability({}).supported
        False
    """
    if not server_headers:
        return MppCapability(supported=False, mode="none", reason="no MPP advertisement")

    for header_name in _MPP_VERSION_HEADERS:
        value = _read_header_ci(server_headers, header_name)
        if value:
            return MppCapability(supported=True, mode="detect-only", spec_version=value)

    challenge = _read_header_ci(server_headers, "www-authenticate")
    if challenge and re.search(r"payment", challenge, re.IGNORECASE):
        return MppCapability(supported=True, mode="detect-only", spec_version=challenge)

    return MppCapability(supported=False, mode="none", reason="no MPP advertisement")


@dataclass
class MppAdapter:
    """The shape of an MPP adapter. ``detect`` is real; ``settle`` is a dead-end
    that always raises ``MPP_NOT_IMPLEMENTED`` until MPP exits preview.
    """

    detect: Callable[..., MppCapability]
    settle: Callable[..., Any]


def create_mpp_seam() -> MppAdapter:
    """Create the MPP capability seam.

    The returned adapter can DETECT MPP advertisements but cannot settle: its
    ``settle`` raises :class:`PayBotApiError` ``MPP_NOT_IMPLEMENTED`` (HTTP 501).
    This keeps the seam's shape real and testable while keeping the unbuilt
    settlement path loud rather than silently broken.

    :returns: An :class:`MppAdapter` whose ``settle`` always raises.

    :example:
        >>> seam = create_mpp_seam()
        >>> seam.detect({"MPP-Version": "x"}).supported
        True
    """

    def _detect(headers: Optional[Dict[str, str]] = None) -> MppCapability:
        return detect_mpp_capability(headers)

    def _settle(*_args: Any, **_kwargs: Any):
        raise PayBotApiError(
            "MPP settlement is not implemented: full MPP support is deferred until "
            "MPP exits preview. The EIP-3009 signing path does not transfer to MPP "
            "(different challenge/credential model); this seam exists to detect and "
            "route, not to pay.",
            "MPP_NOT_IMPLEMENTED",
            501,
        )

    return MppAdapter(detect=_detect, settle=_settle)

File: src/mpp-seam.ts
/**
 * @module mpp-seam
 *
 * Thin **MPP (Merchant/Machine Payments Protocol — Stripe/Tempo) capability
 * seam**. This is deliberately NOT a client.
 *
 * WHY a seam and not a client:
 *   At time of writing, Stripe/Tempo MPP is still in PREVIEW: version-pinned and
 *   bound to the Tempo L1, and it shares NO signing code with our EIP-3009 core.
 *   Our x402 signing path (EIP-3009 `TransferWithAuthorization` over EVM) does
 *   not transfer to MPP — MPP uses a different challenge/credential model. So
 *   building a full MPP client now would mean committing to a moving,
 *   preview-stage target with a separate crypto stack.
 *
 *   Instead, this module provides the *shape* of MPP support: it can DETECT
 *   whether a server advertises MPP (so a caller can negotiate/route), and it
 *   exposes a `settle` entry point that LOUDLY throws `MPP_NOT_IMPLEMENTED`.
 *   When MPP exits preview, the real settlement path slots in behind this seam
 *   without changing the public surface.
 *
 * Dependencies: ./errors (PayBotApiError).
 * Used by: routing/negotiation code that wants to know "can this endpoint do MPP
 * yet?" without paying for it.
 */

import { PayBotApiError } from './errors.js';

/**
 * The result of probing a server for MPP support.
 *
 * - `mode: 'none'`        — no MPP advertisement seen; not supported.
 * - `mode: 'detect-only'` — MPP advertised; we can detect/route but NOT settle.
 * - `mode: 'preview'`     — reserved for a future preview-grade settlement path.
 */
export interface MppCapability {
  /** Whether any MPP advertisement was detected. */
  supported: boolean;
  /** Capability mode. Today only `'none'` and `'detect-only'` are produced. */
  mode: 'none' | 'detect-only' | 'preview';
  /** The advertised spec/version string, when the server provided one. */
  specVersion?: string;
  /** Human-readable reason, populated when `supported` is `false`. */
  reason?: string;
}

/**
 * Header names that, when present, indicate a server advertises MPP. Compared
 * case-insensitively. `WWW-Authenticate` only counts when its value mentions a
 * payment challenge (see {@link detectMppCapability}).
 */
const MPP_VERSION_HEADERS = ['mpp-version', 'stripe-version'] as const;

/**
 * Read a header value by name, case-insensitively, from a plain record.
 *
 * @param headers - The header record.
 * @param name - Header name to look up (any casing).
 * @returns The value, or `undefined` when absent.
 */
function readHeaderCaseInsensitive(
  headers: Record<string, string>,
  name: string,
): string | undefined {
  const target = name.toLowerCase();
  for (const key of Object.keys(headers)) {
    if (key.toLowerCase() === target) {
      return headers[key];
    }
  }
  return undefined;
}

/**
 * Detect whether response headers advertise MPP support.
 *
 * Detection is purely passive — this NEVER signs or settles. It inspects, case
 * insensitively, for either:
 *   - an `MPP-Version` / `Stripe-Version` header (value becomes `specVersion`), or
 *   - a `WWW-Authenticate` challenge whose value mentions `Payment` (an MPP-style
 *     402 challenge), in which case the challenge value becomes `specVersion`.
 *
 * @param serverHeaders - Response headers to inspect. Omitted/empty → not supported.
 * @returns `{ supported: true, mode: 'detect-only', specVersion }` when MPP is
 *          advertised; otherwise `{ supported: false, mode: 'none', reason }`.
 *
 * @example
 *   detectMppCapability({ 'MPP-Version': '2024-09-preview' });
 *   // { supported: true, mode: 'detect-only', specVersion: '2024-09-preview' }
 *   detectMppCapability({});
 *   // { supported: false, mode: 'none', reason: 'no MPP advertisement' }
 */
export function detectMppCapability(
  serverHeaders?: Record<string, string>,
): MppCapability {
  if (!serverHeaders || Object.keys(serverHeaders).length === 0) {
    return { supported: false, mode: 'none', reason: 'no MPP advertisement' };
  }

  for (const headerName of MPP_VERSION_HEADERS) {
    const value = readHeaderCaseInsensitive(serverHeaders, headerName);
    if (value && value.length > 0) {
      return { supported: true, mode: 'detect-only', specVersion: value };
    }
  }

  // WWW-Authenticate: Payment ... → MPP-style 402 challenge.
  const challenge = readHeaderCaseInsensitive(serverHeaders, 'www-authenticate');
  if (challenge && /payment/i.test(challenge)) {
    return { supported: true, mode: 'detect-only', specVersion: challenge };
  }

  return { supported: false, mode: 'none', reason: 'no MPP advertisement' };
}

/**
 * The shape of an MPP adapter. `detect` is real; `settle` is intentionally a
 * dead-end (`never`) until MPP exits preview — see {@link createMppSeam}.
 */
export interface MppAdapter {
  /** Detect MPP support from response headers. */
  detect(headers?: Record<string, string>): MppCapability;
  /**
   * Settle via MPP. NOT implemented — always throws `MPP_NOT_IMPLEMENTED`. The
   * signature is `(...args) => never` so the unbuilt path is loud and typed.
   */
  settle(...args: unknown[]): never;
}

/**
 * Create the MPP capability seam.
 *
 * The returned adapter can DETECT MPP advertisements but cannot settle: its
 * `settle` throws {@link PayBotApiError} `MPP_NOT_IMPLEMENTED` (HTTP 501) with a
 * message pointing to the deferral rationale. This keeps the seam's shape real
 * and testable while keeping the unbuilt settlement path loud rather than
 * silently broken.
 *
 * @returns An {@link MppAdapter} whose `settle` always throws.
 *
 * @example
 *   const seam = createMppSeam();
 *   seam.detect({ 'MPP-Version': 'x' }).supported; // true
 *   seam.settle();                                 // throws MPP_NOT_IMPLEMENTED (501)
 */
export function createMppSeam(): MppAdapter {
  return {
    detect(headers?: Record<string, string>): MppCapability {
      return detectMppCapability(headers);
    },
    settle(..._args: unknown[]): never {
      throw new PayBotApiError(
        'MPP settlement is not implemented: full MPP support is deferred until ' +
          'MPP exits preview. The EIP-3009 signing path does not transfer to ' +
          'MPP (different challenge/credential model); this seam exists to ' +
          'detect and route, not to pay.',
        'MPP_NOT_IMPLEMENTED',
        501,
      );
    },
  };
}
